Buttplug.Server/Bluetooth/Devices/Lovense.cs

```diff
@@ -196,36 +195,101 @@ internal class Lovense : ButtplugBluetoothDevice
+            {
+                MsgFuncs.Add(typeof(RotateCmd), new ButtplugDeviceWrapper(HandleRotateCmd, new MessageAttributes() { FeatureCount = 1 }));
+            }
         }
 
         private async Task<ButtplugMessage> HandleStopDeviceCmd(ButtplugDeviceMessage aMsg)
         {
             BpLogger.Debug("Stopping Device " + Name);
+
+            if (friendlyNames[Interface.Name] == "Nora")
+            {
+                await Interface.WriteValue(aMsg.Id,
+                    Info.Characteristics[(uint)LovenseRev1BluetoothInfo.Chrs.Tx],
+                    Encoding.ASCII.GetBytes($"Rotate:0;"));
+            }
+
             return await HandleSingleMotorVibrateCmd(new SingleMotorVibrateCmd(aMsg.DeviceIndex, 0, aMsg.Id));
         }
 
         private async Task<ButtplugMessage> HandleSingleMotorVibrateCmd(ButtplugDeviceMessage aMsg)
         {
             var cmdMsg = aMsg as SingleMotorVibrateCmd;
-            if (cmdMsg is null)
+            var cmdMsg2 = aMsg as VibrateCmd;
+            if (cmdMsg is null && cmdMsg2 is null)
             {
                 return BpLogger.LogErrorMsg(aMsg.Id, Error.ErrorClass.ERROR_DEVICE, "Wrong Handler");
             }
 
+            if (cmdMsg != null)
+            {
+                _vibratorSpeeds[0] = cmdMsg.Speed;
+            }
+            else
+            {
+                foreach (var vi in cmdMsg2.Speeds)
+                {
+                    if (vi.Index == 0)
+                    {
+                        _vibratorSpeeds[0] = vi.Speed;
+                    }
+                }
+            }
+
             // While there are 3 lovense revs right now, all of the characteristic arrays are the same.
             return await Interface.WriteValue(aMsg.Id,
                 Info.Characteristics[(uint)LovenseRev1BluetoothInfo.Chrs.Tx],
-                Encoding.ASCII.GetBytes($"Vibrate:{(int)(cmdMsg.Speed * 20)};"));
+                Encoding.ASCII.GetBytes($"Vibrate:{(int)(_vibratorSpeeds[0] * 20)};"));
+        }
+
+        private async Task<ButtplugMessage> HandleRotateCmd(ButtplugDeviceMessage aMsg)
+        {
+            var cmdMsg = aMsg as RotateCmd;
+            if (cmdMsg is null)
+            {
+                return BpLogger.LogErrorMsg(aMsg.Id, Error.ErrorClass.ERROR_DEVICE, "Wrong Handler");
+            }
+
+            var dirChange = false;
+            foreach (var vi in cmdMsg.Speeds)
+            {
+                if (vi.Index == 0)
+                {
+                    _rotateSpeed = vi.Speed;
+                    dirChange = _clockwise != vi.Clockwise;
+                }
+            }
+
+            if (dirChange)
+            {
+                _clockwise = !_clockwise;
+                await Interface.WriteValue(aMsg.Id,
+                   Info.Characteristics[(uint)LovenseRev1BluetoothInfo.Chrs.Tx],
+                   Encoding.ASCII.GetBytes($"RotateChange;"));
+            }
+
+            return await Interface.WriteValue(aMsg.Id,

```

Comment:
Writing GATT commands is takes a non-trivial amount of time. If we're gonna batch here, we may want to check what our previous vibration speed set was, and if it's the same, don't send it again. That means we can reverse directions in a rotate message without having to incur 2 GATT transfers.

Reply:
I should be doing that for all the GATT sends, but I do worry about packet drop.
I might see if I can get the GATT write to confirm if it actually did work, and if not, don't update the state.

Whilst I'm at it, an option for "do not skip this" might be handy, for stop device signals, etc.

Reply:
If there's already a GATT write in flight, it should throw an exception, right? Isn't that the only thing we care about?